### DISCUSSION TITLE
Question: Installing from CRAN in pixi?

### DISCUSSION THREAD

**Discussion by kaedonkers** [roaldarbol/rpix#12] [Q&A] [answered]
Hi roaldarbol,

Awesome project, much needed in the data science world!

I have been frustrated by the need to install R packages through a conda channel for a couple of years now. It feels like installing directly from CRAN would be a much more robust and clean solution, however I have never found anyone working on it. Given that this can be done with Python packages from pip (in either conda or pixi), I feel like it should be possible with R packages from CRAN.

Is this something you hope to achieve in this project? Do you have any thoughts on how installing directly from CRAN into a pixi environment (or indeed conda environment) might be done?

**Comment by roaldarbol** [comment, marked as answer]
Hey @kaedonkers!

Thanks for sharing the excitement! This package simply serves as an interface for pixi, so all the packaging is dealt with on the pixi side. But we are discussing all these things; I personally feel a robust conda packaging solution is needed, which the team is building in `rattler-build`. Alternatively they would have to work on something like porting renv. They did something similarlybfor pip, but since it was a lot of work and uv cane along it made more sense to utilize their work. 

It’s almost possible to package all the r-universe packages (which I believe includes all CRAN packages), it just need to be automated - however it’s their goal. You can follow the conversations here and give a thumbs up to the issue to let them know it should be prioritised.

{rpix} is simply going to be wrapping pixi CLI commands so we can install conda packages the way we R users normally install packages (in the R console), so there’s none of the “fun” stuff happening here, simply hoping to enable a good workflow akin to {rix} does for Nix.

**Comment by kaedonkers** [comment]
Great, thanks for the update! I agree that a conda-like solution for R packages is needed. I have tried `renv` but it doesn't quite deliver the polyglot solutions I need. I see you have blogged about using `renv`, any opinions on how that fits in?

Exciting to hear that the prefix team are working on automating the R-to-conda package pipeline, that will solve a lot!

**Comment by roaldarbol** [comment]
Yeah, it's the cross-language solutions I'm missing too. `renv` works well for R projects. I use it in pretty much all the projects I work on, and it has some nice features such as the one I blogged about. So yeah, great tool! For myself, I couldn't wait for a conda solution before getting on with my work, so I see it as an intermediate solution until we have great conda tooling. Consider joining their discord channel, would be great to have some more R voices to discuss with. :-)

**Comment by andystopia** [comment]
I can say that I would appreciate a CRAN native solution which can use conda dependencies would be so sweet. Sort of like how uv and pixi can interop. Being able to use packages with tricky builds like sf, with a lockfile on it's conda dependencies, like GDAL, if that's possible, would be incredible. 

The CRAN has a few architectural limitations which make creating a good package manager hard (all as far as I know, please call out mistakes)
 - there is no way to answer the query: "give me the dependency bounds of ggplot 3.5.0" without downloading the entire package, pulling out it's description file and parsing it. This is extremely slow if a package has 50+ versions (which some do!). You send 50MB across the wire to read a 40kb file. 
 - GitHub maintains a cran mirror, which is *much* faster than CRAN, but it's hard to rely on that sort of thing, considering Microsoft owns GitHub and Microsoft killed the MRAN (which used to host historical versions of the CRAN), but it could be a work-around.
- Packages don't have a structured format for system dependencies, they either bundle them (so compilers would be necessary), or expect you to figure it out.
- autobrew is a project which tries to remedy some of this, but it depends on homebrew, and I would like a cross platform solution, but could be a good reference.
- R makes no forward / backward compatibility guarantees. You cannot build once per R 4.x.x and be good for all R 4.x.x. Every new version requires a new build, which is a huge quadratic explosion in things to build / host every time an R version is released.


What I've been working on (I'm busy with college lol):
 - A from source, by default (like Cargo), with a binary substitution option disabled by default in the manifest, for CRAN packages. 
   - this lets me makes sure I can set up a project correctly so it builds, and then can send it out to a team and they will be able to either pull down the binaries or build from source themself.
 - A solution which doesn't require teammates to have to install gfortran and C++ compilers. I know R recommends them. Almost nobody I've worked with has either installed. I would like these to be provided. R is a scripting language and I am often working with people who have little interest with system package management.
 - Possible minimum version selection.
 
I think an package index which is maintained would be super slick. Similar to what cargo does. This could actually enable strong version solving. 

Overall, I'm really hopeful for the strengths that conda/pixi can bring to the ecosystem. There's a great deal of packages out there which do major versions by releasing PackageName2, or which pull binaries from GitHub releases for each release to avoid builds on client machines. I believe that pixi will be a huge productivity / reproducibility boon in the scientific computing world :D
